src/materials/LinearSpring.C

```diff
@@ -0,0 +1,247 @@
+  _original_global_to_local_rotation(0, 0) = x_orientation(0);
+  _original_global_to_local_rotation(0, 1) = x_orientation(1);
+  _original_global_to_local_rotation(1, 0) = y_orientation(0);
+  _original_global_to_local_rotation(1, 1) = y_orientation(1);
+  _original_global_to_local_rotation(0, 2) = x_orientation(2);
+  _original_global_to_local_rotation(1, 2) = y_orientation(2);
+  _original_global_to_local_rotation(2, 0) = z_orientation(0);
+  _original_global_to_local_rotation(2, 1) = z_orientation(1);
+  _original_global_to_local_rotation(2, 2) = z_orientation(2);
+
+  // _qp = 0
+  computeDeformations();
+
+  // computing spring forces
+  computeForces();
+
+  // computing spring stiffness matrix
+  computeStiffnessMatrix();
+}
+
+void
+LinearSpring::computeTotalRotation()
+{
+  _qp = 0;
+
+  // Currently this forumation is limited to small deformations in the spring,
+  // namely, it is assumed that there are no rigid body rotations in the spring,
+  // and that the total rotation matrix from global to local coordinates
+  // (calculated below) remains the same as the one at t = 0 throughout the
+  // duration of the simulation.
+  _total_global_to_local_rotation[_qp] = _original_global_to_local_rotation;
+}
+
+void
+LinearSpring::computeDeformations()
+{
+  // fetch the two end nodes for _current_elem
+  std::vector<Node *> node;
+  for (unsigned int i = 0; i < 2; ++i)
+    node.push_back(_current_elem->get_node(i));
+
+  // Fetch the solution for the two end nodes at time t
+  NonlinearSystemBase & nonlinear_sys = _fe_problem.getNonlinearSystemBase();
+  const NumericVector<Number> & sol = *nonlinear_sys.currentSolution();
+
+  // Calculating global displacements and rotations at the end nodes
+  RealVectorValue global_disp0(3, 0.0); // node 0
+  RealVectorValue global_disp1(3, 0.0); // node 1
+  RealVectorValue global_rot0(3, 0.0);  // node 0
+  RealVectorValue global_rot1(3, 0.0);  // node 1
+  for (unsigned int i = 0; i < _ndisp; ++i)
+  {
+    global_disp0(i) = sol(node[0]->dof_number(nonlinear_sys.number(), _disp_num[i], 0));
+    global_disp1(i) = sol(node[1]->dof_number(nonlinear_sys.number(), _disp_num[i], 0));
+    global_rot0(i) = sol(node[0]->dof_number(nonlinear_sys.number(), _rot_num[i], 0));
+    global_rot1(i) = sol(node[1]->dof_number(nonlinear_sys.number(), _rot_num[i], 0));
+  }
+
+  // Convert spring nodal displacements and rotations from global coordinate system to local
+  // coordinate system First, compute total rotation
+  computeTotalRotation();
+  RealVectorValue local_disp0(_total_global_to_local_rotation[_qp] * global_disp0);
+  RealVectorValue local_disp1(_total_global_to_local_rotation[_qp] * global_disp1);
+  RealVectorValue local_rot0(_total_global_to_local_rotation[_qp] * global_rot0);
+  RealVectorValue local_rot1(_total_global_to_local_rotation[_qp] * global_rot1);
+
+  // Calculating spring deformations and rotations. Deformations and rotations
```

Comment:
Calculating spring deformations and rotations along the local coordinate system of the spring.